src/ripple/core/SQLInterface.h

```diff
@@ -0,0 +1,247 @@
+
+    typedef std::pair<LedgerIndex, uint8_t> DatabaseIndex;
+
+    static std::string
+    tableName(TableType type)
+    {
+        switch (type)
+        {
+            case ledgers:
+                return "Ledgers";
+            case transactions:
+                return "Transactions";
+            case accountTransactions:
+                return "AccountTransactions";
+            case lastTable:
+                static_assert(lastTable == 3);
+        }
+        return "Unknown";
+    }
+
+    struct RowsMinMax {
+        int rows;
+        int min;
+        int max;
+    };
+
+    struct Hashes {
+        uint256 ledgerHash;
+        uint256 parentHash;
+    };
+
+    static std::unique_ptr<SQLInterface>
+    init(Application& app, Config const& config);
+
+    virtual bool
+    makeLedgerDBs(
+        Config const& config,
+        DatabaseCon::Setup const& setup,
+        DatabaseCon::CheckpointerSetup const& checkpointerSetup) = 0;
+
+    virtual void
+    addDBs(LedgerIndex shardIndex, DatabaseCon* tx, DatabaseCon* lgr) = 0;
+
+    virtual void
+    removeDBs(LedgerIndex shardIndex) = 0;
+
+    virtual ~SQLInterface() = default;
+
+    virtual boost::optional<LedgerIndex>
+    getMinLedgerSeq(DatabaseType dbtype, TableType type) = 0;
+
+    virtual boost::optional<LedgerIndex>
+    getMaxLedgerSeq(DatabaseType dbtype, TableType type) = 0;
+
+    virtual void
+    deleteByLedgerSeq(
+        DatabaseType dbtype,
+        TableType type,
+        LedgerIndex ledgerSeq) = 0;
+
+    virtual void
+    deleteBeforeLedgerSeq(
+        DatabaseType dbtype,
+        TableType type,
+        LedgerIndex ledgerSeq) = 0;
+
+    virtual int
+    getRows(DatabaseType dbtype, TableType type) = 0;
+
+    virtual struct RowsMinMax
+    getRowsMinMax(DatabaseType dbtype, TableType type) = 0;
+
+    virtual bool
+    saveValidatedLedger(
+        Application& app,
+        std::shared_ptr<Ledger const> const& ledger,
+        bool current) = 0;
+
+    virtual std::optional<SQLLedgerInfo>
+    loadLedgerInfoByIndex(
+        LedgerIndex ledgerSeq,
+        beast::Journal j) = 0;
+
+    virtual std::optional<SQLLedgerInfo>
+    loadLedgerInfoByIndexSorted(
+        bool ascendSort,
+        beast::Journal j) = 0;
+
+    virtual std::optional<SQLLedgerInfo>
+    loadLedgerInfoByIndexLimitedSorted(
+        LedgerIndex ledgerFirstIndex,
+        bool ascendSort,
+        beast::Journal j) = 0;
+
+    virtual std::optional<SQLLedgerInfo>
+    loadLedgerInfoByHash(
+        uint256 const& ledgerHash,
+        beast::Journal j) = 0;
+
+    virtual uint256
+    getHashByIndex(LedgerIndex ledgerIndex) = 0;
+
+    virtual std::optional<Hashes>
+    getHashesByIndex(
+        LedgerIndex ledgerIndex,
+        beast::Journal j) = 0;
+
+    virtual std::map<LedgerIndex, Hashes>
+    getHashesByIndex(
+        LedgerIndex minSeq,
+        LedgerIndex maxSeq,
+        beast::Journal j) = 0;
+
+    virtual Json::Value
```

Comment:
Let's not return JSON. `std::vector<STTx>` or `std::vector<std::shared_ptr<STTx>>` would be better

Reply:
Replaced by std::vector<std::shared_ptr<Transaction>>. It contain more information than STTx, for example, status.